zorch/lnp/challenge.py

```
@@ -0,0 +1,187 @@
+import numpy as np
+from lattice_frx.roots import galois_map
+from lattice_frx.sampler import uniform_bytes_needed, uniform_from_bytes
+
+
+def _require_params(d: int, kappa: int, eta: int, k: int, fail_prob: float) -> None:
+    """One validation for both public entry points. The failure modes stay
+    `ValueError` throughout — every case is a wrong parameter value."""
+    if d < 4 or d & (d - 1):
+        raise ValueError(
+            f"challenge: degree must be a power of two >= 4 to carry the "
+            f"σ₋₁-invariant shape, got {d!r}"
```

Ruff (0.16.1)

[warning] String contains ambiguous `σ` (GREEK SMALL LETTER SIGMA). Did you mean `o` (LATIN SMALL LETTER O)?

(RUF001)